### PR TITLE
Adds Support for Managing AWS Load Balancer Type

### PR DESCRIPTION
**Pull request by danehans**
Extends and implements the Contour API for managing AWS load balancer types. Supports Classic ELB and NLB.

Fixes https://github.com/projectcontour/contour-operator/issues/187

xref: https://github.com/projectcontour/contour-operator/pull/336

cc: @BostjanBozic

### CODE REVIEW

**Review by youngnick — APPROVED** [review]

File: internal/objects/service/service.go [hunk 1]
```diff
@@ -83,7 +81,7 @@ var (
 	//  https://kubernetes.io/docs/concepts/services-networking/service/#internal-load-balancer
 	//
 	LbAnnotations = map[operatorv1alpha1.LoadBalancerProviderType]map[string]string{
-		awsProviderType: {
+		operatorv1alpha1.AWSLoadBalancerProvider: {
 			awsLbBackendProtoAnnotation: "tcp",
```
Comment (BostjanBozic) [inline internal/objects/service/service.go:85]: I believe this annotation should only be included if Classic type of Load Balancer is specified (https://projectcontour.io/guides/deploy-aws-nlb/)

File: internal/objects/service/service.go [hunk 2]
```diff
@@ -94,14 +92,14 @@ var (
 	//  https://kubernetes.io/docs/concepts/services-networking/service/#internal-load-balancer
 	//
 	InternalLBAnnotations = map[operatorv1alpha1.LoadBalancerProviderType]map[string]string{
-		awsProviderType: {
+		operatorv1alpha1.AWSLoadBalancerProvider: {
 			awsInternalLBAnnotation: "0.0.0.0/0",
```
Comment (BostjanBozic) [inline internal/objects/service/service.go:96]: I believe this should be changed from `0.0.0.0/0` to `"true"`. Ref: https://kubernetes.io/docs/concepts/services-networking/service/#internal-load-balancer

File: internal/objects/service/service.go [hunk 3]
```diff
@@ -262,6 +260,14 @@ func DesiredEnvoyService(contour *operatorv1alpha1.Contour) *corev1.Service {
 		svc.Spec.Type = corev1.ServiceTypeLoadBalancer
 		provider := contour.Spec.NetworkPublishing.Envoy.LoadBalancer.ProviderParameters.Type
 		lbAnnotations := LbAnnotations[provider]
+		// Add the NLB annotation to the map if specified by AWS provider parameters.
+		if provider == operatorv1alpha1.AWSLoadBalancerProvider &&
+			contour.Spec.NetworkPublishing.Envoy.Type == operatorv1alpha1.LoadBalancerServicePublishingType &&
```
Comment (BostjanBozic) [inline internal/objects/service/service.go:265]: Isn't this already tested with above `switch` clause?

File: internal/objects/service/service.go [hunk 4]
```diff
@@ -262,6 +260,14 @@ func DesiredEnvoyService(contour *operatorv1alpha1.Contour) *corev1.Service {
...
+			contour.Spec.NetworkPublishing.Envoy.LoadBalancer.ProviderParameters.Type == operatorv1alpha1.AWSLoadBalancerProvider &&
```
Comment (BostjanBozic) [inline internal/objects/service/service.go:266]: Isn't this already tested in `if` clause above?

File: internal/objects/service/service.go [hunk 5]
```diff
@@ -262,6 +260,14 @@ func DesiredEnvoyService(contour *operatorv1alpha1.Contour) *corev1.Service {
...
+			contour.Spec.NetworkPublishing.Envoy.LoadBalancer.ProviderParameters.AWS != nil &&
```
Comment (BostjanBozic) [inline internal/objects/service/service.go:267]: Would this not be obsolete, since next condition already includes `!= nil` condition in it?

**Review by BostjanBozic — CHANGES_REQUESTED** [review]
@danehans I checked this, looks good. Only thing that I believe has to change is now `aws-load-balancer-backend-protol` annotation is handled, as it should be implemented only if `Classic` load balancer is used: https://projectcontour.io/guides/deploy-aws-nlb/

File: internal/objects/service/service.go (hunk 5 above)
Comment (danehans) [reply internal/objects/service/service.go:267]: AWS provider params is an optional field. Checking to make sure it's not nil and the param type is NLB.

**Comment by danehans** [comment]
@BostjanBozic commit `6c4d753` resolves your feedback above, PTAL.

**Review by BostjanBozic — APPROVED** [review]

**Comment by BostjanBozic** [comment]
@danehans LGTM

**Review by sunjayBhatia — APPROVED** [review]

**Pull request merged**